compiler/src/iree/compiler/InputConversion/MHLO/ConvertCollectiveOps.cpp

```diff
@@ -0,0 +1,411 @@
+    Value gatherInput = op.getOperand();
+    DenseIntElementsAttr permutationAttr;
+
+    if (allGatherDim != 0) {
+      SmallVector<int64_t> permutation =
+          llvm::to_vector(llvm::seq<int64_t>(0, gatherResultShape.size()));
+      std::swap(permutation[0], permutation[allGatherDim]);
+      permutationAttr = rewriter.getI64VectorAttr(permutation);
+      std::swap(gatherInputShape[0], gatherInputShape[allGatherDim]);
+      std::swap(gatherResultShape[0], gatherResultShape[allGatherDim]);
+      // transpose the input
+      gatherInput =
+          rewriter
+              .create<mhlo::TransposeOp>(
+                  loc, RankedTensorType::get(gatherInputShape, elemType),
+                  gatherInput, permutationAttr)
+              .getResult();
+    }
+
+    // Create an empty tensor for the result
+    Value target = rewriter.create<tensor::EmptyOp>(
+        loc, gatherResultShape, resultType.getElementType());
+    Value gatherResult =
+        rewriter
+            .create<IREE::Flow::AllGatherOp>(op.getLoc(), elementTypeAttr,
+                                             target, gatherInput, channel)
+            .getResult();
+
+    if (allGatherDim != 0) {
+      gatherResult = rewriter
+                         .create<mhlo::TransposeOp>(
+                             loc, resultType, gatherResult, permutationAttr)
+                         .getResult();
+    }
+
+    rewriter.replaceOp(op, gatherResult);
+    return success();
+  }
+};
+
+struct AllReduceOpConversion : public OpConversionPattern<mhlo::AllReduceOp> {
+  using OpConversionPattern<mhlo::AllReduceOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::AllReduceOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+
+    if (!op.getUseGlobalDeviceIds())
+      return rewriter.notifyMatchFailure(op, "must use global device IDs");
+
+    // Check there is only one group in the replica_groups
+    ShapedType replicaGroupType = op.getReplicaGroups().getType();
+    if (replicaGroupType.getRank() != 2 ||
+        replicaGroupType.getDimSize(0) != 1) {
+      return rewriter.notifyMatchFailure(op,
+                                         "must have a single replica group");
+    }
+
+    // Only single elementwise op is supported.
+    Block &block = op.getComputation().front();
+
+    if (block.empty() || llvm::hasSingleElement(block) ||
+        std::next(block.begin(), 2) != block.end())
+      return rewriter.notifyMatchFailure(op, "must have two ops in the block");
+
+    if (block.getNumArguments() != 2)
+      return rewriter.notifyMatchFailure(op, "must have two block args");
+
+    Operation &op1 = block.front();
+    Operation &op2 = *(++block.begin());
+
+    if (op1.getNumResults() != 1 ||
+        !op1.hasTrait<::mlir::OpTrait::Elementwise>())
+      return rewriter.notifyMatchFailure(op, "must have elementwise trait");
```

Comment:
any multi-line ifs should be wrapped with braces (here and above/below) - llvm is pretty lax on this but in the IREE codebase we follow the google guideline there: https://google.github.io/styleguide/cppguide.html#Conditionals

the easiest rule to remember is to always add braces - it's always safer and unless hyper optimizing to save one vertical whitespace line is not any less readable. part of this is because avoiding the braces makes refactoring and automatic reformatting much more complex: a line that may have just fit because the error message string had one word instead of another may suddenly not with a global find/replace, and then all places either end up inconsistent or need to be fixed. or if someone comes along and decides they need to do one more thing instead of a diff of 1 line they now need to change 3 to add the braces. It's the same reason having trailing commas on initializer lists is useful.